### PR TITLE
[BLB] Implement Eluge the shoreless sea and expand SpellsCostReductionControllerEffect functionality

### PR DESCRIPTION
**Pull request by jimga150**
#11853

### CODE REVIEW

**Comment by jimga150** [comment]
I checked with the verifycardstest (setting the oracle test errors to fail) that this hasn't changed any existing oracle tests.

File: Mage/src/main/java/mage/abilities/effects/common/cost/SpellsCostReductionControllerEffect.java [hunk 1]
```diff
@@ -32,15 +33,25 @@ public class SpellsCostReductionControllerEffect extends CostModificationEffectI
     private final DynamicValue amount;
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/effects/common/cost/SpellsCostReductionControllerEffect.java:33]: Add docs like:
* `// number of times to apply cost reduction`

File: Mage/src/main/java/mage/abilities/effects/common/cost/SpellsCostReductionControllerEffect.java [hunk 2]
```diff
@@ -32,15 +33,25 @@ public class SpellsCostReductionControllerEffect extends CostModificationEffectI
...
     private final boolean upTo;
     private ManaCosts<ManaCost> manaCostsToReduce = null;
+    private final boolean convertToGeneric;
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/effects/common/cost/SpellsCostReductionControllerEffect.java:36]: Add docs like:
* `// if no more colored mana then allow to reduce generic cost, see CardUtil.adjustCost`

File: Mage/src/main/java/mage/abilities/effects/common/cost/SpellsCostReductionControllerEffect.java [hunk 3]
```diff
@@ -71,14 +82,20 @@ protected SpellsCostReductionControllerEffect(final SpellsCostReductionControlle
...
+        this.convertToGeneric = effect.convertToGeneric;
+        this.staticText = effect.staticText;
     }
 
     @Override
     public boolean apply(Game game, Ability source, Ability abilityToModify) {
+        int reductionAmount = this.amount.calculate(game, source, this);
         if (manaCostsToReduce != null) {
-            CardUtil.adjustCost((SpellAbility) abilityToModify, manaCostsToReduce, false);
+            ManaCosts<ManaCost> calculatedManaCostsToReduce = new ManaCostsImpl<>();;
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/effects/common/cost/SpellsCostReductionControllerEffect.java:93]: There is `upTo` mode to choose any amount to reduce. Current code:
* colored reduce + upTo -- non implemented
* generic reduce + upTo -- implemented

I recommends to add runtime error in constructor like
``` java
        // runtime check
        if (this.manaCostsToReduce != null && this.upTo) {
            throw new IllegalStateException("Wrong code usage: colored cost reduce un-support "up to" mode");
        }
```

**Review by JayDi85 — APPROVED** [review]

File: Mage/src/main/java/mage/abilities/effects/common/cost/SpellsCostReductionControllerEffect.java (hunk 3 above)
Comment (jimga150) [reply Mage/src/main/java/mage/abilities/effects/common/cost/SpellsCostReductionControllerEffect.java:93]: No constructs set both of those parameters, so with the current code that can't happen.

**Pull request merged**